httpcore/_async/http_proxy.py

```diff
@@ -140,47 +140,52 @@ def __init__(
         connection = await self._get_connection_from_pool(origin)
 
         if connection is None:
-            connection = AsyncHTTPConnection(
-                origin=origin, http2=False, ssl_context=self._ssl_context,
+            # First, create a connection to the proxy server
+            proxy_connection = AsyncHTTPConnection(
+                origin=self.proxy_origin, http2=False, ssl_context=self._ssl_context,
             )
             async with self._thread_lock:
                 self._connections.setdefault(origin, set())
-                self._connections[origin].add(connection)
+                self._connections[origin].add(proxy_connection)
 
-            # Establish the connection by issuing a CONNECT request...
+            # Issue a CONNECT request...
 
             # CONNECT www.example.org:80 HTTP/1.1
             # [proxy-headers]
             target = b"%b:%d" % (url[1], url[2])
             connect_url = self.proxy_origin + (target,)
-            connect_headers = self.proxy_headers
-            proxy_response = await connection.request(
+            connect_headers = self.proxy_headers or [(b"host", self.proxy_origin[1])]
+            proxy_response = await proxy_connection.request(
                 b"CONNECT", connect_url, headers=connect_headers, timeout=timeout
             )
             proxy_status_code = proxy_response[1]
             proxy_reason_phrase = proxy_response[2]
             proxy_stream = proxy_response[4]
 
-            # Ingest any request body.
-            await read_body(proxy_stream)
+            # Read the response data without closing the socket
+            _ = b"".join([chunk async for chunk in proxy_stream])
 
             # If the proxy responds with an error, then drop the connection
             # from the pool, and raise an exception.
             if proxy_status_code < 200 or proxy_status_code > 299:
-                async with self._thread_lock:
-                    self._connections[connection.origin].remove(connection)
-                    if not self._connections[connection.origin]:
-                        del self._connections[connection.origin]
                 msg = "%d %s" % (proxy_status_code, proxy_reason_phrase.decode("ascii"))
                 raise ProxyError(msg)
 
-            # Upgrade to TLS.
-            await connection.start_tls(target, timeout)
+            # Create a new connection to the target
+            connection = AsyncHTTPConnection(
+                origin=origin,
+                http2=False,
+                ssl_context=self._ssl_context,
+                socket=proxy_connection.socket,
+            )
+            async with self._thread_lock:
+                self._connections[origin].remove(proxy_connection)
+                self._connections[origin].add(connection)
```

Comment:
Indeed, we are not calling TLS anymore because I left it out for a subsequent PR 🙂 

The process is actually quite similar, it's just that `httpcore` has an additional level of indirection in the `AsyncHTTPConnection` that makes things harder.

`AsyncHTTPConnection` uses the backend to create a TCP socket and creates one of `AsyncHTTP11Connection` and `AsyncHTTTP2Connection`. Crucially it also keeps track of the _state_ of the `h11` connection, which at this point is at `PAUSED` and must be abandoned, making reusing the connection by switching the `origin` impossible. So we need to create a new one to "reset" the logic, but reusing the socket which was not possible before...

... and you just posted a comment saying the exact same thing 😄 

I'll leave this for posterity 👍

Reply:
Okay, yes, I understand now.

So, interestingly if I don't use the `.socket` technique, and just create a new connection and _then_ call `.start_tls()` on it, it will work, i.e. I will get a 404 response from the Squid proxy.

So I assume the "keep the proxy socket open" part is required for _HTTPS_ to work, but not necessarily to fix the bug in #54, right?

Reply:
I'm wondering if we can have this PR fix HTTP-only proxies first, making us _temporarily_ in a "we don't do HTTPS request proxying yet", and _then_ add everything we need to make HTTPS work?

Reply:
(It might be that I'm out of my mind and `.start_tls()` is actually required for tunneling to work, I'm just not sure about that part.)

Reply:
> So, interestingly if I don't use the .socket technique, and just create a new connection and then call .start_tls() on it, it will work, i.e. I will get a 404 response from the Squid proxy.

So `start_tls` is not required for tunneling to work, AFAIK the key is to use the same socket used in the CONNECT request. Wouldn't creating a new connection would create a new socket? Not sure why you're getting that 404, it's hard to say what's happening without looking at the changes 🙂 

> I'm wondering if we can have this PR fix HTTP-only proxies first, making us temporarily in a "we don't do HTTPS request proxying yet", and then add everything we need to make HTTPS work?

Yeah, absolutely. Making HTTPS requests work is literally an if statement to call `start_tls` and bunch of testing fixtures. This PR is the bulk of the changes 🙂

Reply:
> Wouldn't creating a new connection also create a new socket?

Yes currently it would (`._connect()` opens a socket).

> why you're getting that 404

No idea, actually. I was requesting google.com on HTTPS. When I switched to localhost on HTTP I started getting 503. I think I just don't know enough about Squid to make things work, might try with mitmproxy more often. :rofl: